File: builder/builder.go
// Package builder provides functionality for resolving a full node.Node graph.
package builder

import (
	"github.com/ankeesler/btool/log"
	"github.com/ankeesler/btool/node"
	"github.com/pkg/errors"
)

//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . Currenter

// Currenter tells whether a node.Node needs to be resolved.
type Currenter interface {
	Current(*node.Node) (bool, error)
}

//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . Callback

// Callback is an interface that clients can use to be notified about resolutions
// when they happen. It also tells the clients whether or not the node was up
// to date.
type Callback interface {
	OnResolve(n *node.Node, current bool)
}

// Builder can Build() a full node.Node graph.
type Builder struct {
	dryRun bool
	c      Currenter
	cb     Callback
}

// New creates a new Builder.
func New(dryRun bool, c Currenter, cb Callback) *Builder {
	return &Builder{
		dryRun: dryRun,
		c:      c,
		cb:     cb,
	}
}

func (b *Builder) Build(n *node.Node) error {
	return b.build(n, make(map[*node.Node]bool))
}

func (b *Builder) build(n *node.Node, built map[*node.Node]bool) error {
	if built[n] {
		return nil
	}

	log.Debugf("building %s", n.Name)

	for _, dN := range n.Dependencies {
		log.Debugf("building dependency %s", dN.Name)
		if err := b.build(dN, built); err != nil {
			return errors.Wrap(err, "resolve "+dN.Name)
		}
	}

	current, err := b.c.Current(n)
	if err != nil {
		return errors.Wrap(err, "current")
	}

	b.cb.OnResolve(n, current)

	if n.Resolver != nil {
		if !current {
			log.Debugf("resolving %s", n.Name)
			if !b.dryRun {
				if err := n.Resolver.Resolve(n); err != nil {
					return errors.Wrap(err, "really resolve "+n.Name)
				}
			}
		}
	}

	built[n] = true

	return nil
}

File: cc/app/builder/builder.cc
#include "app/builder/builder.h"

#include <string>

#include "err.h"
#include "log.h"
#include "node/node.h"

namespace btool::app::builder {

bool Builder::Build(const ::btool::node::Node &node, std::string *ret_err) {
  bool success = true;

  node.Visit([this, ret_err, &success](const ::btool::node::Node *n) {
    if (success) {
      bool ret_current;
      std::string current_ret_err;
      auto success = c_->Current(*n, &ret_current, &current_ret_err);
      DEBUGS() << "builer visiting " << n->name() << ", current: "
               << (success ? (ret_current ? "true" : "false") : current_ret_err)
               << ", resolver = "
               << (n->resolver() == nullptr ? "null" : "something")
               << std::endl;
      if (!success) {
        success = false;
        *ret_err = ::btool::WrapErr(current_ret_err, "current");
      } else if (!ret_current && n->resolver() != nullptr) {
        auto err = n->resolver()->Resolve(*n);
        if (err) {
          success = false;
          *ret_err = err.Msg();
        }
      }
    }
  });

  return success;
}

};  // namespace btool::app::builder
